src/ArraySetops.chpl

```diff
@@ -114,30 +122,149 @@ module ArraySetops
+            const localizedPullLocalCount = pullLocalCount;
+            const aDom = a.localSubdomain();
+            const bDom = b.localSubdomain();
+
+            // the ternaries allow me to declare the next variables as const
+            // info for pushing to previous locale
+            const numSendBack = if loc == Locales[0] then 0 else localizedPullLocalCount[here.id - 1];
+            const sendBack = if loc == Locales[0] then (false, false) else localizedPullLocalFlag[here.id - 1];
+            // info for fetching from next locale
+            const numPullLocal = localizedPullLocalCount[here.id];
+            const pullLocal = pullLocalFlag[here.id];
+
+            // skip any indices that were pulled to previous locale
+            const aLow = if sendBack[0] then (aDom.first + numSendBack) else aDom.first;
+            const bLow = if sendBack[1] then (bDom.first + numSendBack) else bDom.first;
+            const aSlice = aLow..<(aDom.last + 1);
+            const bSlice = bLow..<(bDom.last + 1);
+
+            var tmp: [0..<(aDom.size + bDom.size - numSendBack + numPullLocal)] (t, t2);
+
+            // write the local part of a (minus any skipped values) into tmp
+            tmp[0..<aSlice.size] = [(key,val) in zip(a.localSlice[aSlice], aVal.localSlice[aSlice])] (key,val);
+            // write the local part of b (minus any skipped values) into tmp
+            tmp[aSlice.size..#bSlice.size] = [(key,val) in zip(b.localSlice[bSlice], bVal.localSlice[bSlice])] (key,val);
+
+            // add any indices that we need to pull to this locale
+            // this not a local slice, but it should (hopefully) take advantage of bulk transfer
+            if pullLocal[0] {
+              const pullFromASlice = (aDom.last+1)..#numPullLocal;
+              tmp[(aSlice.size+bSlice.size)..#(numPullLocal+1)] = [(key,val) in zip(a[pullFromASlice], aVal[pullFromASlice])] (key,val);
+            }
+            else if pullLocal[1] {
+              const pullFromBSlice = (bDom.last+1)..#numPullLocal;
+              tmp[(aSlice.size+bSlice.size)..#(numPullLocal+1)] = [(key,val) in zip(b[pullFromBSlice], bVal[pullFromBSlice])] (key,val);
+            }
+
+            // run chpl's builtin parallel radix sort for local arrays with a comparator defined in
+            // RadixSortLSD.chpl which sorts on the key
+            twoArrayRadixSort(tmp, new KeysRanksComparator());
+
+            const writeToResultSlice = (segs[here.id]+numSendBack)..#tmp.size;
+            sortedIdx[writeToResultSlice] = [(key, val) in tmp] key;
+            permutedVals[writeToResultSlice] = [(key, val) in tmp] val;
+          }
+        }
+      }
+      else {
+        var perm = makeDistArray(allocSize, int);
+        forall (s, p, sp) in zip(sortedIdx, perm, radixSortLSD(concatArrays(idx1, idx2, ordered=false))) {
+          (s, p) = sp;
+        }
+        const vals = concatArrays(val1, val2, ordered=false);
```

Comment:
yah so it's prob worth a comment, so we are concatenating the indices array and then the values arrays. This will make sure that indices will line up with their associated values

Reply:
I should have looked more closely.  Thanks for clarifying.